### ISSUE TITLE
using decorators to avoid code replication

### ISSUE THREAD

**Issue by andrea-bistacchi** [gecos-lab/PZero#115]
Avoiding code replication is generally achieved by using classes. However, this solution is not optimal when dealing with functions, simply defined as:

```
def(self, some_arguments)
    ...
```

In this case we might consider using decorators following this pattern:

```
def a_decorator(func):

    def wrapper(*args, **kwargs):
        # do something here, the same for all wrapped functions
        ...
        # the wrapped function goes here
        return func(*args, **kwargs)
        # do something here, the same for all wrapped functions
        ...

    return wrapper


@a_decorator
def a_wrapped_function(some_argument):
    # do something here, specific to the wrapped function
    ...
    return some_output
```

I will test this to unify freezing and unfreezing the GUI in functions of `two_d_lines.py`.

**Comment by andrea-bistacchi** [comment]
In 92dc29b9 I introduced the `@freeze_gui` decorator.

The decorator itself can be found at:
https://github.com/gecos-lab/PZero/blob/92dc29b91fac7714d6e7908796803b1c44ff01d4/pzero/helpers/helper_functions.py#L286

The first usage of this decorator can be found at:
https://github.com/gecos-lab/PZero/blob/92dc29b91fac7714d6e7908796803b1c44ff01d4/pzero/two_d_lines.py#L208

@r-monti14 please go ahead using the decorator in all functions in [two_d_lines.py](https://github.com/gecos-lab/PZero/blob/class_refactoring_pyside6/pzero/two_d_lines.py) and [three_d_surfaces.py](https://github.com/gecos-lab/PZero/blob/class_refactoring_pyside6/pzero/three_d_surfaces.py), following the same pattern as above.

In addition to avoiding code repetition, this will solve #114 and similar issues.

**Comment by waqashussain117** [comment]
After adding the decorater and removing the functions that are already described in the decorater, the error is still there. The window freezes.

**Comment by r-monti14** [comment]
The decorator works with those tools that do not need a final input to interrupt the function. 
Tools such as "Edit line", "Draw line" and "Split line-point", after being enhanced with the decorator, do not block the window.

**Comment by andrea-bistacchi** [comment]
With 7a98c742 I have updated both `resample_line_distance` and `resample_line_number_points` functions.

@r-monti14 please use these as a reference for all other functions in `two_d_lines.py` and `three_d_surfaces.py`.

We will think about the interactive tools `Edit line`, `Draw line` and `Split line-point` later on.

**Comment by r-monti14** [comment]
With f6d562f I have added the decorators at all the functions in two_d_lines.py (except Draw line, Edit line, Split line-point, Extend line and Measure).

**Comment by andrea-bistacchi** [comment]
Great! Have you tested all of them?

**Comment by andrea-bistacchi** [comment]
Yes thanks. Actually I think it would be a good idea to use freezing everywhere, also in functions used to open or import data.
